posts/improving-the-access-control-of-a-jhipster-application.md

```diff
@@ -0,0 +1,222 @@
+---
+authors:
+- Theo Lebrun
+tags:
+- JHipster
+- Spring
+- Spring Security
+- React
+- Angular
+date: 2019-01-22T15:24:45.000Z
+title: "Improving the access-control of a JHipster application"
+image: https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2019/01/access-control.png
+---
+
+By default, a JHipster application will let any authenticated user the right to perform the four basic functions (or CRUD) on any entities. This behavior is very useful to showcase the application since any user can for example create an entity and update it. However, for a real application in production that kind of behavior is not really appropriate. The application's routes and API need to be restricted so a regular user does not have total freedom on every entities.
+
+In this blog post, I will explain how to improve the access-control of the entities generated by JHipster. That covers the API for the backend and the frontend's routes/UI for both Angular/React. This kind of improvements can be easily achieved and I recommend doing it as soon as possible to avoid having issues when going to production.
+
+# Backend customization
+
+## Spring Security and JHipster authorities
+
+JHipster uses [Spring Security](https://spring.io/projects/spring-security) to secure the application, it is a very mature and robust framework used to secure any Spring-based applications. Spring Security can be easily customized to change the application's authentication and access-control to fulfill any desired requirements. JHipster uses 4 kind of users (**system**, **anonymousUser**, **user** and **admin**) that have one or multiple authorities (**ROLE_ANONYMOUS**, **ROLE_USER** and **ROLE_ADMIN**). More information about the type of users and authorities can be found on the [JHipster security page](https://www.jhipster.tech/security/).
+
+User's access-control can be configured using authorities at two different levels: 
+- Url
+    - **user** can use the entities API
+    - only **admin** can use the management's API
+- HTTP verb
+    - **user** can get an user using the verb `GET`
+    - only **admin** can remove an user using the verb `DELETE`
+
+Thereby, the application's API access-control can be configured at an url level or more specifically at a HTTP verb level. That gives some nice flexibility and allow to achieve any access-control scenarios.
```

Comment:
That gives us some nice flexibility and allows us to configure any access-control policy.